zboxcore/sdk/reader.go

```diff
@@ -0,0 +1,232 @@
+	// in StreamDownload struct
+	BlocksFor10MB = 160
+)
+
+type StreamDownloadOption struct {
+	ContentMode     string
+	AuthTicket      string
+	BlocksPerMarker uint // Number of blocks to download per request
+	VerifyDownload  bool // Verify downloaded data against ValidaitonRoot.
+}
+
+type StreamDownload struct {
+	*DownloadRequest
+	offset   int64
+	open     bool
+	fileSize int64
+}
+
+func (sd *StreamDownload) Close() error {
+	sd.open = false
+	return nil
+}
+
+func (sd *StreamDownload) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset > sd.fileSize {
+			return 0, errors.New(ExceededMaxOffsetValue, "")
+		}
+		sd.offset = offset
+	case io.SeekCurrent:
+		if sd.offset+offset >= sd.fileSize {
+			return 0, errors.New(ExceededMaxOffsetValue, "")
+		}
+		sd.offset += offset
+	case io.SeekEnd:
+		newOffset := sd.fileSize - offset
+		if newOffset < 0 {
+			return 0, errors.New(NegativeOffsetResultantValue, "")
+		}
+		sd.offset = offset
+	default:
+		return 0, errors.New(InvalidWhenceValue,
+			fmt.Sprintf("expected 0, 1 or 2, provided %d", whence))
+	}
+	return sd.offset, nil
+}
+
+// getStartAndEndIndex will return start and end index based on fileSize, offset and wantSize value
+func (sd *StreamDownload) getStartAndEndIndex(wantsize int64) (int64, int64) {
+	sizePerBlobber := (sd.fileSize +
+		int64(sd.datashards) - 1) / int64(sd.datashards) // equivalent to ceil(filesize/datashards)
+
+	totalBlocksPerBlobber := (sizePerBlobber +
+		int64(sd.effectiveBlockSize) - 1) / int64(sd.effectiveBlockSize)
+
+	effectiveChunkSize := sd.effectiveBlockSize * sd.datashards
+	startInd := sd.offset / int64(effectiveChunkSize)
+	endInd := (sd.offset + wantsize + int64(effectiveChunkSize) - 1) / int64(effectiveChunkSize)
+	if endInd > totalBlocksPerBlobber {
+		endInd = totalBlocksPerBlobber
+	}
+	return startInd, endInd
+}
+
+func (sd *StreamDownload) Read(b []byte) (int, error) {
+	if !sd.open {
+		return 0, errors.New("file_closed", "")
+	}
+
+	if sd.offset >= sd.fileSize {
+		return 0, io.EOF
+	}
+
+	wantSize := int64(math.Min(float64(len(b)), float64(sd.fileSize-sd.offset)))
+	if wantSize <= 0 {
+		return 0, ErrInvalidRead
+	}
+
+	startInd, endInd := sd.getStartAndEndIndex(wantSize)
+	var numBlocks int64
+	if sd.numBlocks > 0 {
+		numBlocks = sd.numBlocks
+	} else {
+		numBlocks = endInd - startInd
+		if numBlocks > BlocksFor10MB {
+			numBlocks = BlocksFor10MB
+		}
+	}
+
+	effectiveChunkSize := sd.effectiveBlockSize * sd.datashards
+	n := 0
+	for startInd < endInd {
+		if startInd+numBlocks > endInd {
+			// this numBlocks should not exceed number greater than required data
+			// otherwise `no shard data` error will occur in erasure reconstruction.
+			numBlocks = endInd - startInd
+		}
+
+		data, err := sd.getBlocksData(startInd, numBlocks)
+		if err != nil {
+			return 0, err
+		}
+
+		offset := sd.offset % int64(effectiveChunkSize)
```

Comment:
Not sure what does the mod do here. And seems it could cause data lose. Imagine we request block_1, the offset is 109, and effectiveChunkSize is 100, then we will copy the `data[9:]` in block_1 to `b`. So where's the `data[:9]` going.

Reply:
@peterlimg If client requests with offset 109, then it means client does not want or care about `data[:9]`. So for example, if client already had stored 109 bytes of data then for requesting other data offset would be set to 109. Since data is requested in block level(100 bytes in above example), 109 bytes means client has stored 100 bytes from block_0 and 9 bytes from block_1, so now after getting block_1, it should skip `data[:9]` otherwise data will be corrupted.